TOY-56 다이어리 메인화면 MVVM + Kotlin 변환

pby2017 opened the pull request:
코드 리뷰 부탁드립니다.
- 메인 화면 MVVM + Kotlin 변환
- 달력으로 화면 이동 구현
- 다이어리 시간 최신 순서 우선으로 정렬

pby2017 committed:
TOY-56 다이어리 메인 MVVM + Kotlin 변환

- 달력으로 화면 이동 구현
- 다이어리 시간 최신 순서 우선으로 정렬

hyejin830 commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/BaseActivity.kt:
```diff
@@ -42,6 +48,12 @@ open class BaseActivity : AppCompatActivity() {
         progressBar.visibility = View.GONE
     }
 
+    fun buildCalendarDialog(calendar: Calendar, listener: DatePickerDialog.OnDateSetListener): YearMonthPickerDialog {
```
다이어리 메인에서만 사용할 것 같은데 BaseActivity에 작성한 이유가 따로 있을까요?!

pby2017 replied:
Dialog 라는 이름 때문에 아직 없지만 다른 곳에서도 사용할 수 있을 것이라고 생각해서 BaseActivity에 작성했는데,
현재는 달력 다이얼로그가 필요한 클래스 파일은 다이어리 메인 밖에 없어서 
다이어리 메인으로 옮기도록 하겠습니다.

hyejin830 commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/DiariesActivity.kt:
```diff
@@ -0,0 +1,89 @@
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_diaries)
```
이 부분은 없어도 되지 않을까요?!

pby2017 replied:
네,
없어도 되는 코드입니다. 삭제하겠습니다.

hyejin830 commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/DiariesAdapter.kt:
```diff
@@ -0,0 +1,77 @@
+        init {
+            itemView.setOnClickListener {
+                val intent = Intent(context, DiaryDetailActivity::class.java)
+                intent.putExtra("DIARY_IDX", diaries[adapterPosition].idx)
```
이 부분 전에 쓰던 방식과 동일하게 하시는 건 어떠신가요?!
```
val intent = Intent(context, DiaryDetailActivity::class.java).apply{
       putExtra("DIARY_IDX", diaries[adapterPosition].idx)
}          
```

pby2017 replied:
네,
반영했습니다.

KimSoungRyoul commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/DiariesActivity.kt:
```diff
@@ -0,0 +1,89 @@
+import kr.co.yogiyo.rookiephotoapp.databinding.ActivityDiariesBinding
+import kr.co.yogiyo.rookiephotoapp.diary.DiaryEditActivity
+import kr.co.yogiyo.rookiephotoapp.settings.SettingsActivity
+import java.util.*
```
필요한 아이들만 import해주는게 좋을 것 같아요
### Avoid import * !!!
https://stackoverflow.com/questions/22000507/java-avoid-repeated-import-not-inherit-import

pby2017 replied:
네,
꼭 확인하겠습니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/BaseActivity.kt:
```diff
@@ -26,7 +32,7 @@ open class BaseActivity : AppCompatActivity() {
     }
 
-    fun addProgressBarInto(layout: RelativeLayout) {
+    fun addProgressBarInto(layout: ViewGroup) {
```
이거 그냥 로딩뷰 보여주는건가요?

bonchangkoo replied:
로딩뷰를 위한거라면 로딩뷰를 DialogFragment로 처리하도록 해요~

pby2017 replied:
네, 
DialogFragment 로 구현해서 처리하겠습니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/DiariesActivity.kt:
```diff
@@ -0,0 +1,89 @@
+import kr.co.yogiyo.rookiephotoapp.settings.SettingsActivity
+import java.util.*
+
+class DiariesActivity : BaseActivity(), DatePickerDialog.OnDateSetListener {
```
뷰모델 쓰는데 listener가 왜 있는거에요?

pby2017 replied:
날짜를 선택하는 다이얼로그 PickerDialog 의 날짜 선택 결과를 되돌려 받아 해당 날짜의 페이지로 이동하기 위해 Listener를 사용했습니다. 

무명 객체로 Listener를 구현해서 Listener implements를 제거하겠습니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/DiariesAdapter.kt:
```diff
@@ -0,0 +1,77 @@
+
+class DiariesAdapter(val context: Context, var diaries: List<Diary>) : RecyclerView.Adapter<DiariesAdapter.DiariesViewHolder>() {
+
+    private var hourMinuteFormat = SimpleDateFormat("hh:mm", Locale.getDefault())
```
var 이어야 하는 이유가 있어요? 그리고 by lazy로 써도 될거 같아요.

pby2017 replied:
아뇨 없습니다.
`val`과 `by lazy` 사용해서 수정하겠습니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/DiariesAdapter.kt:
```diff
@@ -0,0 +1,77 @@
+class DiariesAdapter(val context: Context, var diaries: List<Diary>) : RecyclerView.Adapter<DiariesAdapter.DiariesViewHolder>() {
+
+    private var hourMinuteFormat = SimpleDateFormat("hh:mm", Locale.getDefault())
+    private var dayFormat = SimpleDateFormat("dd일", Locale.getDefault())
```
위에랑 마찬가지요.

pby2017 replied:
네,
`val`과 `by lazy` 사용해서 수정하겠습니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/DiariesAdapter.kt:
```diff
@@ -0,0 +1,77 @@
+            time = date
+        }
+
+        diariesViewHolder.dayText.text = dayFormat.format(calendar.time)
```
diariesViewHolder.run { } 으로 묶어서 써보세요.

pby2017 replied:
네,
반영하겠습니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/DiariesAdapter.kt:
```diff
@@ -0,0 +1,77 @@
+
+        init {
+            itemView.setOnClickListener {
+                val intent = Intent(context, DiaryDetailActivity::class.java)
```
.apply { } 적용해보세요~

pby2017 replied:
네,
반영하겠습니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/DiariesFragment.kt:
```diff
@@ -0,0 +1,99 @@
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        diariesViewModel = ViewModelProviders.of((context as DiariesActivity)).get(DiariesViewModel::class.java)
```
context as DiariesActivity -> this

```
    @NonNull
    @MainThread
    public static ViewModelProvider of(@NonNull Fragment fragment, @Nullable Factory factory) {
        Application application = checkApplication(checkActivity(fragment));
        if (factory == null) {
            factory = ViewModelProvider.AndroidViewModelFactory.getInstance(application);
        }
        return new ViewModelProvider(ViewModelStores.of(fragment), factory);
    }
```

모르는 함수 호출할 때는 어떻게 쓰는지 먼저 확인부터하는 습관을 들이세요.

pby2017 replied:
네, 
해당 ViewModel은 Activity에서 인스턴스화한 ViewModel을 Fragment에서도 사용하기 위해 activity로 부터 전달받은 context를 캐스팅해서 사용했었습니다. 

this (fragment)를 인자로 전달했을 때 Activity ViewModel의 hashCode와 각 Fragment들의 ViewModel의 hashCode가 모두 달라서 동일한 hashCode의 ViewModel을 사용하기 위해 context를 활용했었습니다.

해당 부분은 getActivity!! 로 수정하겠습니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/DiariesFragment.kt:
```diff
@@ -0,0 +1,99 @@
+        val root = inflater.inflate(R.layout.fragment_diaries, container, false)
+
+        diariesRecyclerView = root.findViewById<RecyclerView>(R.id.recycler_diaries).apply {
+            val linearLayoutManager = LinearLayoutManager(context)
```
```suggestion
            layoutManager = LinearLayoutManager(context)
            adapter = DiariesAdapter(context, ArrayList())
```

불필요한 코드는 줄입시다...

pby2017 replied:
네, 
해당 변수는 하단의 `addItemDecoration` 설정에서 인자로 전달하고 있어서 따로 변수로 생성했습니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/DiariesFragment.kt:
```diff
@@ -0,0 +1,99 @@
+        private const val POSITION = "position"
+
+        fun newInstance(context: DiariesActivity, position: Int): Fragment {
+            val bundle = Bundle()
```
```suggestion
            val bundle = Bundle().apply { ... }
```

apply 써보도록 하세요.

pby2017 replied:
네,
반영하겠습니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/DiariesViewModel.kt:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    private fun getCalendarFromPosition(position: Int): Calendar {
+        val calendar = GregorianCalendar()
```
apply { ... } 적용해보세요.

pby2017 replied:
네,
반영하겠습니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/DiariesViewModel.kt:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    private fun howFarFromBase(year: Int, month: Int): Int {
+        val calendar = Calendar.getInstance()
```
apply { ... } 적용해보세요.

pby2017 replied:
네,
반영하겠습니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/YearMonthPickerDialog.kt:
```diff
@@ -0,0 +1,79 @@
+        val view = activity!!.layoutInflater.inflate(R.layout.picker_year_month, null)
+
+        val builder = AlertDialog.Builder(context).apply {
+            yearPicker = view.findViewById<NumberPicker>(R.id.picker_year).apply {
```
코틀린에서는 findViewById를 쓸일이 없어요...

 view.picker_year.apply { }

pby2017 replied:
네,
반영하겠습니다.

`view.` 없이 picker_year 만 사용했을 때 null이 표시되서 
`view.` 를 사용하면서 자바 코드처럼 작성했었습니다.
숙지하겠습니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/YearMonthPickerDialog.kt:
```diff
@@ -0,0 +1,79 @@
+                }
+            }
+
+            monthPicker = view.findViewById<NumberPicker>(R.id.picker_month).apply {
```
코틀린에서는 findViewById를 쓸일이 없어요. (2)

pby2017 replied:
네,
반영하겠습니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/YearMonthPickerDialog.kt:
```diff
@@ -0,0 +1,79 @@
+                }
+            }
+
+            confirmButton = view.findViewById<Button>(R.id.btn_confirm).apply {
```
코틀린에서는 findViewById를 쓸일이 없어요. (3)

pby2017 replied:
네,
반영하겠습니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/YearMonthPickerDialog.kt:
```diff
@@ -0,0 +1,79 @@
+                    dismiss()
+                }
+            }
+            cancelButton = view.findViewById<Button>(R.id.btn_cancel).apply {
```
코틀린에서는 findViewById를 쓸일이 없어요. (4)

pby2017 replied:
네,
반영하겠습니다.

bonchangkoo reviewed (changes requested):
리뷰 확인 부탁드려요~

pby2017 committed:
TOY-56 피드백 반영

pby2017 committed:
TOY-56 충돌 해결

pby2017 committed:
TOY-56 충돌 해결

Merge branch 'develop' into TOY-56-diary-main-mvvm-kotlin

# Conflicts:
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/camera/capture/PreviewActivity.kt
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/db/LocalDiaryViewModel.kt
#	app/src/main/res/values/strings.xml

pby2017 commented:
@bonchangkoo 님, @hyejin830 님, @KimSoungRyoul 님,

피드백 반영 완료했습니다.
확인 부탁드립니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/DiariesFragment.kt:
```diff
@@ -5,93 +5,92 @@ import android.os.Bundle
+        val view = inflater.inflate(R.layout.fragment_diaries, container, false)
 
-        diariesRecyclerView = root.findViewById<RecyclerView>(R.id.recycler_diaries).apply {
+        view.recycler_diaries.apply {
```
apply -> run 이 더 적당한거 같아요~

apply, run, let 등등의 용법을 한번 찾아보세요~

pby2017 replied:
네,
피드백 반영하겠습니다.

apply(T.() -> Unit)
객체 내부를 수정하고 객체를 반환할 때,

also((T) -> Unit)
 객체 내부를 수정하지 않거나 사용하지 않고 객체를 반환할 때,

let((T) -> R)
null이 아닌 객체 자신을 사용해서 코드를 실행하고 결과가 필요없거나 마지막 줄을 반환할 때,

with(receiver: T, block: T.() -> R)
run()과 비슷하면서 객체를 명시적으로 나타낼 때,

run(() -> R)
run(T.() -> R)
객체 없이 코드를 묶거나, null이 아닌 객체 내부를 수정하고 결과가 필요없거나 마지막 줄을 반환할 때

사용합니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/util/LoadingDialogFragment.kt:
```diff
@@ -0,0 +1,20 @@
+        return AlertDialog.Builder(context)
+                .setView(View.inflate(context, R.layout.dialog_fragment_loading, null))
+                .create().apply {
+                    window!!.setBackgroundDrawableResource(android.R.color.transparent)
```
!! (NonNull) 이 아닌 경우도 있을 거 같아요.

pby2017 replied:
네,
!! 대신 
?.run 을 사용해서 null이 아닐 때만 코드를 실행하도록 묶었습니다.

bonchangkoo reviewed (commented):
코멘트 확인 부탁드려요~

pby2017 committed:
TOY-56 피드백 반영

bonchangkoo reviewed (approved):
코드 확인 했습니다.

pby2017 committed:
TOY-56 충돌해결

pby2017 committed:
TOY-56 Merge branch 'develop' into TOY-56-diary-main-mvvm-kotlin

# Conflicts:
#	app/build.gradle
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/BaseActivity.kt
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/Constants.java
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/MainActivity.java
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/DiaryEditActivity.java
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/db/LocalDiaryViewModel.kt
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/DiariesAdapter.java
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/DiariesFragment.java
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/settings/BackupRestoreDialogFragment.java
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/settings/SignDialogFragment.java
#	app/src/main/res/layout/dialog_fragment_backup_restore.xml

pby2017 committed:
TOY-56 색변경, dismiss() 할 때 dialog show() 유무로 발생하던 문제 해결

pby2017 committed:
TOY-56 충돌 해결

Merge branch 'develop' into TOY-56-diary-main-mvvm-kotlin

# Conflicts:
#	app/src/main/res/values/colors.xml
#	app/src/main/res/values/styles.xml
